operator/src/simulation/controller.rs

```diff
@@ -0,0 +1,384 @@
+        source: kube::Error,
+    },
+}
+
+/// Start a controller for the Simulation CRD.
+pub async fn run() {
+    let k_client = Client::try_default().await.unwrap();
+    let context = Arc::new(Context::new(k_client.clone(), HttpRpcClient));
+
+    // Add api for other resources, ie ceramic nodes
+    let networks: Api<Network> = Api::all(k_client.clone());
+    let simulations: Api<Simulation> = Api::all(k_client.clone());
+    let namespaces: Api<Namespace> = Api::all(k_client.clone());
+    let services = Api::<Service>::all(k_client.clone());
+    let config_maps = Api::<ConfigMap>::all(k_client.clone());
+    let jobs = Api::<Job>::all(k_client.clone());
+    let pods = Api::<Pod>::all(k_client.clone());
+
+    Controller::new(simulations.clone(), Config::default())
+        .owns(
+            networks,
+            watcher::Config::default().labels(MANAGED_BY_LABEL_SELECTOR),
+        )
+        .owns(
+            namespaces,
+            watcher::Config::default().labels(MANAGED_BY_LABEL_SELECTOR),
+        )
+        .owns(
+            services,
+            watcher::Config::default().labels(MANAGED_BY_LABEL_SELECTOR),
+        )
+        .owns(
+            config_maps,
+            watcher::Config::default().labels(MANAGED_BY_LABEL_SELECTOR),
+        )
+        .owns(
+            jobs,
+            watcher::Config::default().labels(MANAGED_BY_LABEL_SELECTOR),
+        )
+        .owns(
+            pods,
+            watcher::Config::default().labels(MANAGED_BY_LABEL_SELECTOR),
+        )
+        .run(reconcile, on_error, context)
+        .for_each(|rec_res| async move {
+            match rec_res {
+                Ok((simulation, _)) => {
+                    debug!(simulation.name, "reconcile success");
+                }
+                Err(err) => {
+                    error!(?err, "reconcile error")
+                }
+            }
+        })
+        .await;
+}
+
+/// Perform a reconile pass for the Simulation CRD
+async fn reconcile(simulation: Arc<Simulation>, cx: Arc<Context<impl RpcClient>>) -> Result<Action, Error> {
+    let spec = simulation.spec();
+    debug!(?spec, "reconcile");
+
+    let status = if let Some(status) = &simulation.status {
+        status.clone()
+    } else {
+       SimulationStatus::default()
+    };
+
+    let sim_name = &spec.selector;
+    let net_status = get_network_status(cx.clone(), sim_name).await?;
+    //TODO handle err not found, no matching network
+
+    let net_ready = net_status.ready_replicas == net_status.replicas;
+
+    if !net_ready {
+        debug!("simulation waiting, network not ready");
+        return  Ok(Action::requeue(Duration::from_secs(10)))
+    }
+
+    let num_peers = net_status.ready_replicas;
+    let ns = "keramik-".to_owned() + sim_name;
+
+    apply_jaeger(cx.clone(), &ns, simulation.clone()).await?;
+    apply_prometheus(cx.clone(), &ns, simulation.clone()).await?;
+    apply_opentelemetry(cx.clone(), &ns, simulation.clone()).await?;
```

Comment:
i should probably make sure these are all started/ready before starting workers/managers